Add ability to use GEFS replay ICs

NeilBarton-NOAA opened the pull request:
<!-- PLEASE READ -->
<!-- Any PRs not following this template will be closed -->
<!--
    Please use a short (<60 char), descriptive title above. It should complete the sentence "If merged, this PR will _____". Capitalize the first word and do not end with a period.

    PRs should meet these guidelines:
    - Each PR should address ONE topic and have an associated issue.
    - No hard-coded paths or personal directories.
    - No temporary or backup files should be committed (including logs).
    - Any code that you disabled by being commented out should be removed or reenabled.

    Please delete all these comments before submitting the PR.
-->
# Description
The PR allows the use of ICs from PSL's replay analysis. These replay ICs will be used for GEFS reforecasting and SFS. Two main changes are associated with these updates: (1) replay ICs being valid at 3Z, and (2) the use of warm starts. 

This addresses https://github.com/NOAA-EMC/global-workflow/issues/1838 


# Type of change
<!-- Delete all except one -->
- New feature 

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- GEFS Forecast-only on Hera using replay ICs
- GEFS Forecast-only on Hera using warm start ICs
- GEFS Forecast-only on Hera using cold start ICs
- GFS Forecast-only on Hera CI test

# Checklist
- [X] Any dependent changes have been merged and published
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

NeilBarton-NOAA committed:
running with replay ICs

JessicaMeixner-NOAA committed:
update gdas app hash and turn on C48mx500_3DVarAOWCDA test on hera

JessicaMeixner-NOAA committed:
Bug fix for #2528 from aerorahul

JessicaMeixner-NOAA committed:
add additional update from before

JessicaMeixner-NOAA committed:
remove line that made things "work" but for the wrong reasons

NeilBarton-NOAA committed:
wave updates

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/replay_ics' into SFS_C96X100

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/develop' into replay_ics

NeilBarton-NOAA committed:
small updates for forecast_postdet.sh

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/replay_ics' into SFS_C96X100

JessicaMeixner-NOAA committed:
updates to run with 3 instead of 4 mom restarts
updates to have correct ICs in com for IAU after a cold start first half cycle

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
update CICE restart copying

NeilBarton-NOAA committed:
updates for replay ICS after merge of restart cap

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/develop' into replay_ics

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/replay_ics' into SFS_C96X100

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
update gdas app hash

JessicaMeixner-NOAA committed:
update gdas

NeilBarton-NOAA committed:
updates for GFS CI

NeilBarton-NOAA committed:
addition for GFS test

NeilBarton-NOAA committed:
fixes before PR

WalterKolczynski-NOAA commented on scripts/exgfs_wave_post_pnt.sh:
```diff
@@ -247,8 +247,9 @@ source "${USHgfs}/preamble.sh"
 
-    ${NLN} mod_def.$waveuoutpGRD mod_def.ww3
-    HMS="${cyc}0000"
+    ln -s mod_def.$waveuoutpGRD mod_def.ww3
```
Please retain the use of `${NLN}`

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/replay_ics' into SFS_C96X100

WalterKolczynski-NOAA commented on scripts/exglobal_stage_ic.sh:
```diff
@@ -33,14 +33,25 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
+        ((rc != 0)) && error_message "${src}" "${tgt}" "${rc}"
+        err=$((err + rc))
+    fi
+
```
```suggestion
    prev_atmos_copy_list=(fv_core.res.nc)
    if [[ "${USE_REPLAY_ICS:-"false"}" == "false" ]]; then
      prev_atmos_copy_list+=(coupler.res)
    fi
    for ftype in "${prev_atmos_copy_list[@]}"; do
      src="${BASE_CPLIC}/${CPL_ATMIC:-}/${PDY}${cyc}/${MEMDIR}/atmos/${PDY}.${cyc}0000.${ftype}"
      tgt="${COM_ATMOS_RESTART_PREV}/${PDY}.${cyc}0000.${ftype}"
      ${NCP} "${src}" "${tgt}"
      rc=$?
      ((rc != 0)) && error_message "${src}" "${tgt}" "${rc}"
      err=$((err + rc))
    done
```

WalterKolczynski-NOAA commented on scripts/exglobal_stage_ic.sh:
```diff
@@ -149,7 +171,7 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
   # Stage the WW3 initial conditions to ROTDIR (warm start; TODO: these should be placed in $RUN.$gPDY/$gcyc)
   if [[ "${DO_WAVE:-}" = "YES" ]]; then
-    YMD=${PDY} HH=${cyc} declare_from_tmpl COM_WAVE_RESTART
+    YMD=${gPDY} HH=${gcyc} declare_from_tmpl COM_WAVE_RESTART
```
If this is now going to use the previous cycle's date, the variable name should be changed to `COM_WAVE_RESTART_PREV` to be consistent.

Also, I believe this will break the existing GEFS test.

WalterKolczynski-NOAA commented on scripts/exglobal_stage_ic.sh:
```diff
@@ -72,6 +83,17 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
+  # Atmosphere Perturbation Files (usually used with replay ICS)
+  # Extra zero on MEMDIR ensure we have a number even if the string is empty
+  if (( 0${MEMDIR:3} > 0 )) && [[ "${USE_ATM_PERTURB_FILES:-false}" == "true" ]]; then
+      src="${BASE_CPLIC}/${CPL_OCNIC:-}/${PDY}${cyc}/${MEMDIR}/atmos/${PDY}.${cyc}0000.fv3_perturbation.nc"
```
Why are atmosphere perturbations coming from the ocean IC directory? Doesn't matter right now since they are identical, but this should use `CPL_ATMIC`.

NeilBarton-NOAA replied:
Error in coping and pasting. Updated

WalterKolczynski-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -97,16 +100,28 @@ FV3_postdet() {
         local model_current_time="${current_cycle}"
       fi
+      if [[ "${USE_REPLAY_ICS}" != "true" ]]; then
       rm -f "${DATA}/INPUT/coupler.res"
```
Please indent this block (except the EOF)

WalterKolczynski-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -490,6 +505,7 @@ MOM6_out() {
 
   # Copy MOM6 restarts at the end of the forecast segment to COM for RUN=gfs|gefs
+  if [[ "${USE_REPLAY_ICS}" != "true" ]]; then
   local restart_file
```
Please indent this block.

WalterKolczynski-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -590,6 +607,7 @@ CICE_out() {
 
   # Copy CICE restarts at the end of the forecast segment to COM for RUN=gfs|gefs
+  if [[ "${USE_REPLAY_ICS}" != "true" ]]; then
   local seconds source_file target_file
```
Please indent this block.

WalterKolczynski-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -729,6 +748,7 @@ CMEPS_out() {
 
   # Copy mediator restarts at the end of the forecast segment to COM for RUN=gfs|gefs
+  if [[ "${USE_REPLAY_ICS}" != "true" ]]; then
   echo "Copying mediator restarts for 'RUN=${RUN}' at ${forecast_end_cycle}"
```
Please indent this block.

WalterKolczynski-NOAA commented:
Does this resolve #1838 or is it only part of it? The details were never filled in to that issue to know what needs to be done.

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/c384wcda

NeilBarton-NOAA committed:
Update scripts/exglobal_stage_ic.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

NeilBarton-NOAA committed:
addressing PR comments

NeilBarton-NOAA committed:
edits bugs stage_ics

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/replay_ics' into SFS_C96X100

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
add additional archive switch for ocean/ice DA files for if JEDIOCN is included
--- fixes from Cathy

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/c384wcda' into feature/c384wcda

WalterKolczynski-NOAA reviewed (approved):
Conditionally approved pending successful completion of CI tests. There may also be conflicts with #2545 that will need to be resolved.

JessicaMeixner-NOAA committed:
tab to spaces

WalterKolczynski-NOAA commented:
Didn't notice pynorms was failing. Please remove the trailing whitespace in `gefs_tasks.py`:
```
./workflow/rocoto/gefs_tasks.py:12:43: W291 trailing whitespace
./workflow/rocoto/gefs_tasks.py:13:18: W291 trailing whitespace
```

WalterKolczynski-NOAA reviewed (changes requested):
Didn't notice pynorms was failing. Please remove the trailing whitespace in `gefs_tasks.py`:
```
./workflow/rocoto/gefs_tasks.py:12:43: W291 trailing whitespace
./workflow/rocoto/gefs_tasks.py:13:18: W291 trailing whitespace
```

NeilBarton-NOAA committed:
remove trailing white space

NeilBarton-NOAA commented:
updated 

> Didn't notice pynorms was failing. Please remove the trailing whitespace in `gefs_tasks.py`:
> 
> ```
> ./workflow/rocoto/gefs_tasks.py:12:43: W291 trailing whitespace
> ./workflow/rocoto/gefs_tasks.py:13:18: W291 trailing whitespace
> ```

JessicaMeixner-NOAA committed:
add COPY_FINAL_RESTARTS flag to skip copying restarts at the end

JessicaMeixner-NOAA committed:
saving work with NEXTCYC date, likely will use different solution

JessicaMeixner-NOAA committed:
updating logic for DOIAU related date changes in fewer places
removing extra DOIAU if logic in multiple places... still needs testing

JessicaMeixner-NOAA committed:
create two new variables that define the start time
of the forecast model for current and next cycle and leave other
defintions of variables as is

JessicaMeixner-NOAA committed:
Update ush/forecast_predet.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
use more descriptive variable name sdate -> model start date

JessicaMeixner-NOAA committed:
update cice output ic link

JessicaMeixner-NOAA committed:
update remove data command - suggestion from Rahul

JessicaMeixner-NOAA committed:
Update ush/forecast_predet.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

JessicaMeixner-NOAA committed:
add new parm variable for gefs

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/c384wcda

NeilBarton-NOAA committed:
updates for 1 degree runs

NeilBarton-NOAA committed:
Merge remote-tracking branch 'remotes/JessicaMeixner-NOAA/feature/c384wcda' into SFS_C96X100

NeilBarton-NOAA committed:
edits to stage ICs that start at the end of the update window

NeilBarton-NOAA committed:
replay ICS with merge

NeilBarton-NOAA committed:
get headers in ww3 post output

NeilBarton-NOAA committed:
updates for pycodestyle

NeilBarton-NOAA committed:
more pynorm fixes

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/develop' into SFS_C96X100

NeilBarton-NOAA committed:
fix script errors

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/SFS_C96X100' into replay_ics

NeilBarton-NOAA committed:
adding back MOM override option for merge # Please enter the commit message for
your changes. Lines starting

NeilBarton-NOAA commented:
@WalterKolczynski-NOAA @aerorahul I merged develop into this branch and changed to use files named with the 03 start. In my opinion, it's ready to comments and testing. I tested the GEFS CI and currently testing a GFS forecast.

JessicaMeixner-NOAA commented on parm/config/gefs/config.base:
```diff
@@ -269,7 +269,9 @@ export MEMDIR="mem${ENSMEM}"
 
 # initialize ocean ensemble members with perturbations
 #   if true, only occurs for members greater than zero
+export END_OF_IAU_START=@END_OF_IAU_START@
```
I think this needs to be defined for the GFS configs.

NeilBarton-NOAA replied:
thanks, added

JessicaMeixner-NOAA replied:
@NeilBarton-NOAA I don't see this in parm/config/gfs/config.base, maybe it didn't get pushed?

NeilBarton-NOAA committed:
adding new option to GFS config

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/SFS_C96X100' into replay_ics

NeilBarton-NOAA commented:
@JessicaMeixner-NOAA I added the update to the wrong branch. It is update/merged now. Thanks.

JessicaMeixner-NOAA commented:
> @JessicaMeixner-NOAA I added the update to the wrong branch. It is update/merged now. Thanks.

Thanks @NeilBarton-NOAA!

WalterKolczynski-NOAA commented on parm/config/gefs/config.wave:
```diff
@@ -81,7 +81,11 @@ export WAVEWND_FID=
+    export FHMIN_WAV=3
+else
+    export FHMIN_WAV=0
+fi
```
Note: #2491 is moving this to `config.base`, so we'll need to make sure this change is included no matter what the order of these two PRs.

WalterKolczynski-NOAA commented on ush/wave_outp_spec.sh:
```diff
@@ -31,6 +31,11 @@ source "${USHgfs}/preamble.sh"
+  if [[ "${END_OF_IAU_START}" == "true" ]]; then
+    model_start_date=$(${NDATE} 3 ${CDATE})
+  else
+    model_start_date=${CDATE}
```
We're trying to eliminate `CDATE`, so might as well make this correct from the start.
```suggestion
    model_start_date="${PDY}${cyc}"
```

NeilBarton-NOAA replied:
edited

WalterKolczynski-NOAA commented on workflow/rocoto/gefs_tasks.py:
```diff
@@ -9,48 +10,63 @@ def __init__(self, app_config: AppConfig, cdump: str) -> None:
+            DTG = self._configs['base']['SDATE'] + timedelta(hours=3)
+            DTG_PREFIX = DTG.strftime('%Y%m%d') + '.' + DTG.strftime('%H') + '0000'
+        else:
+            DTG_PREFIX = "@Y@m@d.@H0000"
```
This would fail if used on an experiment of more than one cycle. Note how the non-replay case is using the rocoto `<cyclestr>` formatting (`@Y`, `@m`, etc.). The replay needs to do the same instead of using `SDATE` (see also next comment).
```suggestion
        dtg_prefix = "@Y@m@d.@H0000"
        if self._configs['base']['END_OF_IAU_START']:
            offset = "+03:00:00"
        else:
            offset = None
```

WalterKolczynski-NOAA commented on workflow/rocoto/gefs_tasks.py:
```diff
@@ -9,48 +10,63 @@ def __init__(self, app_config: AppConfig, cdump: str) -> None:
+                            [f'{datatype}.tile{tile}.nc'
+                             for datatype in ['ca_data', 'fv_core.res', 'fv_srf_wnd.res', 'fv_tracer.res', 'phy_data', 'sfc_data']
+                             for tile in range(1, self.n_tiles + 1)]:
+                    data = f"{prefix}/{DTG_PREFIX}.{file}"
+                    dep_dict = {'type': 'data', 'data': data}
+                    deps.append(rocoto.add_dependency(dep_dict))
+                prefix = f"{cpl_ic['BASE_CPLIC']}/{cpl_ic['CPL_ATMIC']}/@Y@m@d@H/mem000/med"
+                data = f"{prefix}/{DTG_PREFIX}.ufs.cpld.cpl.r.nc"
                 dep_dict = {'type': 'data', 'data': data}
```
The `add_dependency()` method can already handle these complicated offset situations (we need to add an offset to the filename but not the directory). We just need to pass an array of data and offsets and they will be combined:
```suggestion
                    data = [prefix, f"{dtg_prefix}.{file}"]
                    dep_dict = {'type': 'data', 'data': data, 'offset': [None, offset]}
                    deps.append(rocoto.add_dependency(dep_dict))
                prefix = f"{cpl_ic['BASE_CPLIC']}/{cpl_ic['CPL_ATMIC']}/@Y@m@d@H/mem000/med"
                data = [prefix, f"{dtg_prefix}.ufs.cpld.cpl.r.nc"]
                dep_dict = {'type': 'data', 'data': data, 'offset': [None, offset]}
```

NeilBarton-NOAA committed:
remove CDATE

NeilBarton-NOAA committed:
Update workflow/rocoto/gefs_tasks.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

NeilBarton-NOAA committed:
Update workflow/rocoto/gefs_tasks.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

NeilBarton-NOAA committed:
Merge branch 'develop' into replay_ics

NeilBarton-NOAA commented:
@WalterKolczynski-NOAA 

There are some errors in the xml file with the updates to workflow/rocoto/gefs_tasks.py
 
Error: Element workflow failed to validate content at

Could you help debug?

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA
> 
> There are some errors in the xml file with the updates to workflow/rocoto/gefs_tasks.py
> 
> Error: Element workflow failed to validate content at
> 
> Could you help debug?

Sorry, was busy much of the day. If you give me a path, I can take a look, but it often means you have a task dependency for a task name that doesn't exist or some other task missing/misspelled issue.

Edit: although since we were mostly mucking with offsets in these changes, it's probably that. I'll need to see what the rendered XML looks like.

WalterKolczynski-NOAA commented:
One error I see for sure is you didn't update the other instances of `dtg_prefix` to be lowercase in `gefs_tasks.py`.

NeilBarton-NOAA committed:
dtg_prefix

NeilBarton-NOAA commented:
@WalterKolczynski-NOAA The above error is associated with something wrong in the xml file, not dtg_prefix. I just uploaded the lower case dtg_prefix so the error can be reproduced.

aerorahul commented on parm/config/gfs/config.base:
```diff
@@ -435,4 +435,7 @@ if [[ ${DO_JEDIATMVAR} = "YES" ]]; then
 fi
 
+# If starting from warm restarts at the end of the IAU window. Used in GEFS when using Replay ICs
+export END_OF_IAU_START="false"
```
A well-rounded way would be to define an `offset` from `current_cycle`.  For replay ICs, the offset is `+3h`.  For cycling/forecast-only with IAU the offset is `-3h`.  If offset is `0h`, the warm starts can be expected to be valid at the `current_cycle`.

aerorahul commented on scripts/exglobal_stage_ic.sh:
```diff
@@ -8,6 +8,14 @@ GDATE=$(date --utc -d "${PDY} ${cyc} - ${assim_freq} hours" +%Y%m%d%H)
 gPDY="${GDATE:0:8}"
 gcyc="${GDATE:8:2}"
 
+END_OF_IAU_START=${END_OF_IAU_START:-"false"}
+if [[ "${END_OF_IAU_START}" == "true" ]]; then
+    RDATE=$(date --utc -d "${PDY} ${cyc} + 3 hours" +%Y%m%d%H)
+    DTG_PREFIX="${RDATE:0:8}.${RDATE:8:2}0000"
+else
+    DTG_PREFIX="${PDY}.${cyc}0000"
+fi
```
formulate in terms of `offset`.

aerorahul commented on workflow/rocoto/gefs_tasks.py:
```diff
@@ -9,48 +10,63 @@ def __init__(self, app_config: AppConfig, cdump: str) -> None:
         deps = []
-
+        dtg_prefix = "@Y@m@d.@H0000"
+        if self._configs['base']['END_OF_IAU_START']:
+            offset = "+03:00:00"
+        else:
+            offset = None
```
This should be handled better. Hard-wiring offsets deep in this code will make this difficult to maintain.
This is a capability that can also be used in GFS if someone wants to start a forecast from the end of the assimilation window e.g. a reanalysis

NeilBarton-NOAA committed:
change variable name for IAU

NeilBarton-NOAA committed:
Merge branch 'NOAA-EMC:develop' into replay_ics

NeilBarton-NOAA committed:
fixes for stage_ic

NeilBarton-NOAA committed:
Merge branch 'replay_ics' of https://github.com/NeilBarton-NOAA/global-workflow into replay_ics

WalterKolczynski-NOAA commented on parm/config/gefs/config.fcst:
```diff
@@ -249,6 +249,9 @@
   export io_layout="1,1"
 fi
 
+if (( ${OFFSET_START_HOUR} != 0 )); then
```
```suggestion
if (( OFFSET_START_HOUR != 0 )); then
```

WalterKolczynski-NOAA commented on scripts/exglobal_stage_ic.sh:
```diff
@@ -33,21 +36,25 @@
-      src="${BASE_CPLIC}/${CPL_ATMIC:-}/${PDY}${cyc}/${MEMDIR}/atmos/${PDY}.${cyc}0000.${ftype}"
-      tgt="${COM_ATMOS_RESTART_PREV}/${PDY}.${cyc}0000.${ftype}"
+    prev_atmos_copy_list=(fv_core.res.nc)
+    if (( ${OFFSET_START_HOUR} == 0 )); then
```
```suggestion
    if (( OFFSET_START_HOUR == 0 )); then
```

WalterKolczynski-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -33,7 +33,10 @@ FV3_postdet() {
     local fv3_restart_files tile_files fv3_restart_file restart_file
-    fv3_restart_files=(coupler.res fv_core.res.nc)
+    fv3_restart_files=(fv_core.res.nc) 
+    if (( ${OFFSET_START_HOUR} == 0 )); then
```
```suggestion
    if (( OFFSET_START_HOUR == 0 )); then
```

WalterKolczynski-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -101,7 +104,9 @@
       # Create a array of increment files
       local inc_files inc_file iaufhrs iaufhr
-      if [[ "${DOIAU}" == "YES" ]]; then
+      if (( ${OFFSET_START_HOUR} != 0 )); then
```
```suggestion
      if (( OFFSET_START_HOUR != 0 )); then
```

WalterKolczynski-NOAA commented on ush/forecast_predet.sh:
```diff
@@ -54,8 +54,11 @@
+  offset_start_begin=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${OFFSET_START_HOUR} hours" +%Y%m%d%H)
   #Define model start date for current_cycle and next_cycle as the time the forecast will start
-  if [[ "${DOIAU:-}" == "YES" ]]; then
+  if (( ${OFFSET_START_HOUR} != 0 )); then
```
```suggestion
  if (( OFFSET_START_HOUR != 0 )); then
```

WalterKolczynski-NOAA reviewed (changes requested):
Sorry for the delay, been a busy kind of week. Also waiting on a CI fix to get in. Just the linter suggestions left, I think.

NeilBarton-NOAA committed:
Update parm/config/gefs/config.fcst

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

NeilBarton-NOAA committed:
Update scripts/exglobal_stage_ic.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

NeilBarton-NOAA committed:
Update ush/forecast_postdet.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

NeilBarton-NOAA committed:
Update ush/forecast_predet.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

NeilBarton-NOAA committed:
Update ush/forecast_postdet.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

WalterKolczynski-NOAA commented:
@NeilBarton-NOAA Please merge in develop and fix any conflicts, and then I think this is ready for CI testing.

aerorahul commented on ush/forecast_det.sh:
```diff
@@ -6,7 +6,7 @@ UFS_det(){
 
   # Determine if the current cycle is a warm start (based on the availability of restarts)
-  if [[ -f "${COM_ATMOS_RESTART_PREV}/${model_start_date_current_cycle:0:8}.${model_start_date_current_cycle:8:2}0000.coupler.res" ]]; then
+  if [[ -f "${COM_MED_RESTART_PREV}/${model_start_date_current_cycle:0:8}.${model_start_date_current_cycle:8:2}0000.ufs.cpld.cpl.r.nc" ]]; then
```
This file is not present when `APP=ATM`, so its not an indicator for warm start.
Besides, one can "warm" start the atm, ocean and ice without this file if set correctly in the ufs_configure (I think).

JessicaMeixner-NOAA replied:
if you set it the mediator to a cold start in the ufs_confugire, you should be able to warm start everything but the mediator.

NeilBarton-NOAA replied:
updated

aerorahul committed:
wip for repay ics

aerorahul committed:
Merge branch 'develop' into feature/replay-1

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/develop' into replay_ics

WalterKolczynski-NOAA reviewed (approved):
Conditionally approved pending successful completion of CI tests.

JessicaMeixner-NOAA commented:
I don't see any issues with the updates made for the wave scripts, which is assuming why my review was asked for.  Also tagging @sbanihash for her awareness of wave script changes.    I didn't make a GFS run to triple check, but visual inspection does not raise any red-flags for any of the updates here for waves.

aerorahul committed:
replay

aerorahul committed:
replay for control member

aerorahul committed:
FHROT is half_window

aerorahul committed:
correct comment

aerorahul committed:
replay w/ cold start ics

WalterKolczynski-NOAA commented:
@NeilBarton-NOAA If you resolve the conflicts, I think this is ready to test.

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/develop' into replay_ics

NeilBarton-NOAA committed:
updates for use of coupler.res and merged with dev

NeilBarton-NOAA commented:
@WalterKolczynski-NOAA updated and merged with dev

WalterKolczynski-NOAA commented:
Looks like multiple issues. One is that the dependency path for wave restart files in stage_ic is missing a slash:

```
  dependencies
    AND is not satisfied
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_ctrl.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile1.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile2.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile3.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile4.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile5.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//gfs_data.tile6.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile1.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile2.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile3.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile4.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile5.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/atmos//sfc_data.tile6.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/ocean/20210323.120000.MOM.res.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/ice/20210323.120000.cice_model.res.nc is available
      /work/noaa/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem000/wave20210323.120000.restart.glo_500 does not exist
```

WalterKolczynski-NOAA commented:
Issue the second is stmp filled on `/work2` again:

```
hercules-login-1 ~/>df /work2/noaa/stmp
Filesystem                                                                            1K-blocks         Used Available Use% Mounted on
172.27.100.53@o2ib:172.27.100.54@o2ib:172.27.100.55@o2ib:172.27.100.56@o2ib:/data2 358612992000 358612992000         0 100% /work2
```

aerorahul committed:
Merge branch 'develop' into feature/replay-1

aerorahul committed:
update for COMIN/COMOUT changes

NeilBarton-NOAA committed:
fix wave stage ic issue

aerorahul committed:
Apply suggestions from code review

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

aerorahul committed:
Merge branch 'develop' into feature/replay-1

aerorahul committed:
Merge branch 'develop' into feature/replay-1

aerorahul committed:
Merge branch 'develop' into replay_ics

aerorahul committed:
Merge branch 'feature/replay-1' into patch/2559-replay1

NeilBarton-NOAA committed:
Merge branch 'NOAA-EMC:develop' into replay_ics

NeilBarton-NOAA committed:
Merge pull request #3 from aerorahul/patch/2559-replay1

patch-1 for 2559

NeilBarton-NOAA committed:
Rahul's suggested changes

NeilBarton-NOAA committed:
remove shell errors

WalterKolczynski-NOAA commented:
@NeilBarton-NOAA is this ready for another try?

NeilBarton-NOAA committed:
update for 3Dvar atmos CI

NeilBarton-NOAA committed:
Merge branch 'replay_ics' of https://github.com/NeilBarton-NOAA/global-workflow into replay_ics

NeilBarton-NOAA committed:
fix accidental delete

NeilBarton-NOAA commented:
@WalterKolczynski-NOAA yes, I tested the 3Dvar ATM CI last night

WalterKolczynski-NOAA committed:
Merge branch 'develop' into replay_ics

WalterKolczynski-NOAA commented:
@NeilBarton-NOAA Setup experiment failed for the GEFS test. Something is wrong in the `stage_ic` task:
```
2024-06-11 18:29:16,135 - DEBUG    - root        : setup_xml.py /scratch1/NCEPDEV/global/CI/2559/RUNTESTS/EXPDIR/C48_S2SWA_gefs_f0410de9
Finalizing initialize
sourcing config.stage_ic
sourcing config.fcst
sourcing config.atmos_products
sourcing config.efcs
sourcing config.atmos_ensstat
sourcing config.waveinit
sourcing config.wavepostsbs
sourcing config.wavepostpnt
sourcing config.oceanice_products
sourcing config.prep_emissions
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/tasks.py", line 239, in get_task
    return getattr(self, task_name, *args, **kwargs)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/gefs_tasks.py", line 16, in stage_ic
    offset = str(self._configs['base']['OFFSET_START_HOUR']).zfill(2) + ":00:00"
                 ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/wxflow/attrdict.py", line 82, in __missing__
    if object.__getattribute__(self, '__frozen'):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'AttrDict' object has no attribute '__frozen'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/CI/2559/gefs//workflow/create_experiment.py", line 101, in <module>
    setup_xml.main(setup_xml_args)
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/setup_xml.py", line 73, in main
    xml = rocoto_xml_factory.create(f'{net}_{mode}', app_config, rocoto_param_dict)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/wxflow/factory.py", line 71, in create
    return self._builders[key](*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/gefs_xml.py", line 14, in __init__
    super().__init__(app_config, rocoto_config)
  File "/scratch1/
```NCEPDEV/global/CI/2559/gefs/workflow/rocoto/workflow_xml.py", line 27, in __init__
    task_list = get_wf_tasks(app_config)
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/workflow_tasks.py", line 21, in get_wf_tasks
    tasks.append(task_obj.get_task(task_name))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2559/gefs/workflow/rocoto/tasks.py", line 241, in get_task
    raise AttributeError(f'"{task_name}" is not a valid task.\n' +
AttributeError: "stage_ic" is not a valid task.
Valid tasks are:
aerosol_init, stage_ic, prep, anal, sfcanl, analcalc, analdiag, arch, cleanup, prepatmiodaobs, atmanlinit, atmanlvar, atmanlfv3inc, atmanlfinal, prepoceanobs, ocnanalprep, ocnanalbmat, ocnanalrun, ocnanalecen, ocnanalchkpt, ocnanalpost, ocnanalvrfy, earc, ecen, echgres, ediag, efcs, eobs, eomg, epos, esfc, eupd, atmensanlinit, atmensanlletkf, atmensanlfv3inc, atmensanlfinal, aeroanlinit, aeroanlrun, aeroanlfinal, prepsnowobs, snowanl, fcst, atmanlupp, atmanlprod, atmupp, goesupp, atmosprod, oceanprod, iceprod, verfozn, verfrad, vminmon, metp, tracker, genesis, genesis_fsu, postsnd, awips_20km_1p0deg, fbwind, gempak, gempakmeta, gempakmetancdc, gempakncdcupapgif, gempakpgrb2spec, npoess_pgrb2_0p5degwaveawipsbulls, waveawipsgridded, wavegempak, waveinit, wavepostbndpnt, wavepostbndpntbll, wavepostpnt, wavepostsbs, waveprep, npoess, mos_stn_prep, mos_grd_prep, mos_ext_stn_prep, mos_ext_grd_prep, mos_stn_fcst, mos_grd_fcst, mos_ext_stn_fcst, mos_ext_grd_fcst, mos_stn_prdgen, mos_grd_prdgen, mos_ext_stn_prdgen, mos_ext_grd_prdgen, mos_wx_prdgen, mos_wx_ext_prdgen
```
(Ignore the "not a valid task" part; that is the error message whenever there is an exception in one of those.)

NeilBarton-NOAA committed:
make sure OFFSET is defined

NeilBarton-NOAA commented:
updated and the GEFS test (and SFS replay test) forecasts are both running on my side

WalterKolczynski-NOAA commented:
> updated and the GEFS test (and SFS replay test) forecasts are both running on my side

K. I'm not starting any more CI on Hera right now until we get some space cleaned up under global, so there might be a slight delay.
